ambry-store/src/main/java/com/github/ambry/store/DiskManager.java

```diff
@@ -570,4 +590,138 @@ private void reportUnrecognizedDirs() {
+     * @return flag to indicate disk healthchecking or not
+     */
+    public boolean isEnabled(){
+      return enabled;
+    }
+
+    /**
+     * Attempts to close the filechannel and delete the file
+     * @param fileChannel the file used for disk healthcheck
+     * @return True/False depending on the close/delete was successful
+     */
+    private boolean deleteFile(AsynchronousFileChannel fileChannel){
+      if(fileChannel.isOpen()){
+        try{
+          fileChannel.close();
+        }catch(IOException e){
+          diskHealthStatus = diskHealthStatus.DELETE_IO_EXCEPTION;
+          logger.error("Exception occurred when deleting a file for disk healthcheck", e);
+          return false;
+        }
+      }
+      return true;
+    }
+
+    /**
+     * Performing Disk HealthCheck Test in 5 steps
+     * 1. Checks if there is sufficient space on disk
+     * 2. Creates the file on the filepath
+     * 3. Writes today's date to disk
+     * 4. Read and compares what was written to disk
+     * 5. Deletes the file
+     */
+    public void diskHealthTest(){
+      if(!enabled){ return;} //ensures disk checks are enabled
+
+      //Creates directory for the file if not already there
+      File dir = new File(heathcheckDir);
+      if (!dir.exists()) dir.mkdirs();
+
+      //Creates and opens the file
+      AsynchronousFileChannel fileChannel;
+      try {
+        Path path = Paths.get(heathcheckDir,"temp");
+        fileChannel =
+            AsynchronousFileChannel.open(path, StandardOpenOption.WRITE, StandardOpenOption.READ,
+                StandardOpenOption.CREATE, StandardOpenOption.DELETE_ON_CLOSE);
+        fileChannel.force(true); //writes to disk immediately
+      }catch(IOException e){
+        diskHealthStatus = DiskHealthStatus.CREATE_IO_EXCEPTION;
+        logger.error("Exception occurred when creating a file for disk healthcheck", e);
+        return;
+      }
+
+      //Finds today's date and stores it
+      String todayDate = new SimpleDateFormat("yyyy-mm-dd hh:mm:ss").format(Calendar.getInstance().getTime());
+      ByteBuffer buffer = ByteBuffer.allocate(todayDate.getBytes().length);
+      buffer.put(todayDate.getBytes());
+      buffer.flip();
+
+      //Write to file in operationTimeoutSeconds seconds
+      Future<Integer> writeOperation = fileChannel.write(buffer, 0);
+      try{
+        writeOperation.get(operationTimeoutSeconds,TimeUnit.SECONDS);
+      }catch(TimeoutException e){
+        writeOperation.cancel(true);
+        diskHealthStatus = DiskHealthStatus.WRITE_TIMEOUT;
+        logger.error("Timeout Exception occurred when writing to a file for disk healthcheck", e);
+        deleteFile(fileChannel);
+        return;
+      }catch(Exception e){
+        writeOperation.cancel(true);
+        diskHealthStatus = DiskHealthStatus.WRITE_EXCEPTION;
+        logger.error("Exception occurred when writing to a file for disk healthcheck", e);
+        deleteFile(fileChannel);
+        return;
+      }
```

Comment:
it seems like we can reuse some of the exception handling code,
```
      Future<Integer> writeOperation = fileChannel.write(buffer, 0);
      Exception ex = null;
      try{
        writeOperation.get(operationTimeoutSeconds,TimeUnit.SECONDS);
      }catch(TimeoutException e){
        ex = e;
        diskHealthStatus = DiskHealthStatus.WRITE_TIMEOUT;
        return;
      }catch(Exception e){
        ex = e;
        diskHealthStatus = DiskHealthStatus.WRITE_EXCEPTION;
        return;
      }
      if (ex != null) {
        writeOperation.cancel(true);
        logger.error("Exception occurred when writing to a file for disk healthcheck", e);
        deleteFile(fileChannel);
        return;
      }
```

Reply:
Great idea. I combined read/write logic since they only differed by a line and refactored that section to make it easier to read